### ISSUE TITLE
Passing user specific parameters/environment variables to spawned single user server?

### ISSUE THREAD

**Issue by dkipping** [jupyterhub/jupyterhub#2258]
(Originally posted as a comment to #1723 , opened up a new issue as the other issue seemed be resolved in a different way)

Our planned scenario:
We are running our own little landing page as a web app and have a form where the user can specify how much memory he needs, a url from which data should be loaded into the environment of the notebook server (should not be persisted, that's why we chose this over mounting volumes.) and a github url to clone a repo from for scripts the user or other users worked on before (not the single-user image, that will be set fix by us. just scripts or notebook files to be loaded).

Is it currently possible to pass custom parameters or environment variables via the jupyterhub api on single user server launch? Because the doc of the /user/{name}/server does not show any parameters or data to be passed except the name of the user.

Or do we need to write our own spawner that takes parameters like this? (still would not make them editable via api but maybe expose them to set them via another service or similar)

Every help appreciated.
Thank you in advance to everyone.

**Comment by dkipping** [comment]
So essentially, the question is: is it possible to provide options for the spawner like offered via the integrated spawner options form: https://jupyterhub.readthedocs.io/en/stable/reference/spawners.html#spawner-options-form

But via api, so that we can run our own options form and do not have to use the integrated one.
Or if that is not possible:
Is it possible to style the optionsform to match the UI our users are used to?

**Comment by minrk** [comment]
Yes, you can definitely do this! If you spawn a server with a POST request with a JSON body, the JSON body becomes the `user_options` dict for the spawner.

I realize this wasn't the best idea to dedicate the whole body for this, so I may deprecate it being at the the top-level and move it down to `options` in the request body.

**Comment by dkipping** [comment]
Great news, thank you. Pointing this out in the documentation is definitely a good idea, so thanks for opening the issue.

**Comment by ankitladhania** [comment]
Hi @minrk,
Can there be further documentation on this?

As in the case of Kube Spawner, `profile` is expected.

**Comment by minrk** [comment]
what user_options are supported will vary by Spawner, so it should be in each spawner's docs what the options should be. No options are supported by the default spawner.

In KubeSpawner, you are right, that 'profile' is the only supported field, and it must be the `display_name` field of the chosen profile.

**Comment by ankitladhania** [comment]
Hi @minrk,
I do understand, I went through the code of KubeSpawner in order to understand this, but still it is not working for me. I have added issue [#1552](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/issues/1552), cause there is no documentation/example to show how it works.
